apps/from_usm_ndarray/usm_ndarray_dlpack.cpp

```diff
@@ -0,0 +1,99 @@
+#include <CL/sycl.hpp>
+#include <cstdint>
+#include <iostream>
+#include "dlpack/dlpack.h"
+
+extern "C" {
+void Give(DLManagedTensor dl_managed_tensor);
+void Finalize();
+void FreeHandle();
+}
+
+namespace {
+
+DLManagedTensor *given = nullptr;
+
+void display(DLManagedTensor mt) {
+    DLTensor tensor = mt.dl_tensor;
+    DLDevice dldev = tensor.device;
+    DLDataType dtype = tensor.dtype;
+    int i, ndim = tensor.ndim;
+    std::cout << "On C++ side" << std::endl;
+    std::cout << "Data: " << std::hex << tensor.data << std::dec << std::endl;
+    std::cout << "Device (" << dldev.device_type << ", " << dldev.device_id << ")" << std::endl;
+    std::cout << "Type description: { 'code': " << static_cast<int>(dtype.code)
+	      << ", 'bits': " << static_cast<int>(dtype.bits)
+	      << ", 'lanes': " << static_cast<int>(dtype.lanes) << " }"  << std::endl;
+    std::cout << "ndim: " << ndim << std::endl;
+    std::cout << "shape: (";
+    {
+	int64_t* sh = tensor.shape;
+	for(int i=0; i < ndim; ++i) {
+	    std::cout << sh[i];
+	    std::cout << ((i+1) < ndim ? ", " : "");
+	}
+    }
+    std::cout << ")" << std::endl;
+
+    std::cout << "strides: (";
+    {
+	int64_t* st = tensor.strides;
+	for(int i=0; i < ndim; ++i) {
+	    std::cout << st[i];
+	    std::cout << ((i+1) < ndim ? ", " : "");
+	}
+    }
+    std::cout << ")" << std::endl;
+
+    sycl::device *d_ptr = nullptr;
+
+    switch(dldev.device_type) {
+    case kDLONEAPI_GPU:
+	d_ptr = new sycl::device(sycl::ONEAPI::filter_selector("gpu:" + std::to_string(dldev.device_id)));
+	break;
+    case kDLONEAPI_CPU:
+	d_ptr = new sycl::device(sycl::ONEAPI::filter_selector("cpu:" + std::to_string(dldev.device_id)));
+	break;
+    case kDLONEAPI_ACCELERATOR:
+	d_ptr = new sycl::device(sycl::ONEAPI::filter_selector("accelerator:" + std::to_string(dldev.device_id)));
+	break;
+    default:
+	throw std::runtime_error("Can not handler these types");
+    }
+
+    sycl::queue q(*d_ptr);   // queue uses default context per root device
+    sycl::context ctxt = q.get_context();
+    delete d_ptr;
+
+    sycl::usm::alloc kind = sycl::get_pointer_type(tensor.data, ctxt);
+    switch (kind) {
+    case sycl::usm::alloc::shared:
+	std::cout << "USM-shared allocation-based array" << std::endl;
+	break;
+    case sycl::usm::alloc::device:
+	std::cout << "USM-device allocation-based array" << std::endl;
+	break;
+    case sycl::usm::alloc::host:
+	std::cout << "USM-host allocation-based array" << std::endl;
+	break;
```

Comment:
Can an allocation made outside of `sycl::usm`, e.g. by numpy, be zero copied into a DLTensor of type `kDLONEAPI_CPU` for use in a sycl program? 

I wonder the same about CUDA/OpenCL buffer allocations, but limit the question to CPU allocations for simplicity.

Reply:
Yes. Host memory can be access from within SYCL kernels through use of ``sycl::buffer``.  

For example, see https://github.com/IntelPython/dpctl/blob/master/examples/pybind11/use_dpctl_syclqueue/pybind11_example.cpp#L61

The function `offloaded_array_mod` works on a regular NumPy array allocated using NumPy's host allocator:

https://github.com/IntelPython/dpctl/blob/master/examples/pybind11/use_dpctl_syclqueue/example.py#L39